Compile/src/Producers/LeptonSFProducer.cc

```diff
@@ -35,15 +38,133 @@ std::string LeptonSFProducer::GetProducerId() const { return "LeptonSFProducer";
 
 void LeptonSFProducer::Init(ZJetSettings const& settings)
 {
-   
-    
+    // Get file
+    LOG(INFO) << this->GetProducerId() << ": Loading file " << m_sffile << ", Histogram "
+              << histoname;
+    TFile file(m_sffile.c_str(), "READONLY");
+    if(!file.Get(histoname.c_str())) {
+        LOG(FATAL) << "Failed to open histogram " << histoname << " from file " << m_sffile;
+    }
+    sfhisto = (TH2F*)file.Get(histoname.c_str());
+    sfhisto->SetDirectory(0);
+    file.Close();
+
+    // use this pointer, so we call child class methods if overwritten.
+    // set correct bools for further processing
+    this->SetEtaAxis2D(histoname);
+    // Use highest pT bin as overflow bin
+    this->SetOverflowPtBin();
 }
 
 void LeptonSFProducer::Produce(ZJetEvent const& event,
-                                 ZJetProduct& product,
-                                 ZJetSettings const& settings) const
+                               ZJetProduct& product,
+                               ZJetSettings const& settings) const
 {
-    
+    LOG(DEBUG) << "\n[" << this->GetProducerId() << "]";
+    if (product.m_zValid) {
+        const auto* lep1 = product.m_originalLeptons.at(product.m_zLeptons.first);
+        const auto* lep2 = product.m_originalLeptons.at(product.m_zLeptons.second);
+        LOG(DEBUG) << "Apply Variation: " << settings.GetLeptonSFVariation();
+        float sf1 = 1 / this->GetScaleFactor(0, *lep1);
+        float sf2 = 1 / this->GetScaleFactor(0, *lep2);
+        product.m_weights["zl1" + weightName] = sf1;
+        product.m_weights["zl2" + weightName] = sf2;
+        product.m_weights["lepton" + weightName] = sf1 * sf2;
```

Comment:
Sure you want to add these Weights to the `m_weights` map? When this producer is configured to run before the Weights-Producer, the produced event weight would be bogus. I would recommend adding only the one weight, which can be considered nominal and add the others to `m_optionalWeights`.

Reply:
Updated

Reply:
Wouldn't that still be a problem for the one weight?

Reply:
What do you mean? Now only the "main" weight is in m_weights, all other weights, including those for individual muons are optional.

Reply:
"When this producer is configured to run before the Weights-Producer, the produced event weight would be bogus" This, I mean.